File: alt_langs/elixir/lib/y2021/Day02.exs
defmodule Day02 do
  def parse_int!(str) do
    case Integer.parse(str) do
      {int, ""} -> int
      _ -> nil
    end
  end

  def parse(input) do
    for line <- String.split(String.trim(input), "\n"),
      line != "" do
      parts = String.split(String.trim(line), " ")
      dir = Enum.at(parts, 0)
      num = Enum.at(parts, 1) |> parse_int!
      case dir do
        "forward" -> {:forward, num}
        "up" -> {:up, num}
        "down" -> {:down, num}
      end
    end
  end

  def part1(input) do
    {horz, depth} = input |> Enum.reduce({0, 0}, fn x, {h, d} -> 
      case x do
        {:forward, num} -> {h + num, d}
        {:down   , num} -> {h, d + num}
        {:up     , num} -> {h, d - num}
      end
    end)
    depth * horz
  end

  def calc_depth([], _, depth, x), do: {x, depth}

  def calc_depth([head | tail], aim, depth, x) do
    case head do
      {:forward, num} -> calc_depth(tail, aim, depth + aim * num, x + num)
      {:down   , num} -> calc_depth(tail, aim + num, depth, x)
      {:up     , num} -> calc_depth(tail, aim - num, depth, x)
    end
  end

  def part2(input) do
    {x, depth} = calc_depth(input, 0, 0, 0)
    x * depth
  end
end

input = File.read("core/shared/src/main/resources/y2021/day02.txt")

data = Day02.parse(input)

IO.puts Day02.part1(data)

IO.puts Day02.part2(data)

File: core/shared/src/main/scala/gay/menkissing/advent/y2021/Day02.scala
package gay.menkissing.advent
package y2021

import cats.syntax.all.*

import scala.annotation.tailrec
import cats.Show

object Day02 extends Problem:
  type Input = List[(Dir, Int)]
  type Output = Int
  def showOutput: Show[Int] = summon

  lazy val input = FileIO.getInput(2021, 2)

  enum Dir:
    case Forward
    case Down
    case Up

  object Dir:
    def parse(str: String): Dir =
      str match
        case "forward" => Dir.Forward
        case "down"    => Dir.Down
        case "up"      => Dir.Up

  def parse(input: String): List[(Dir, Int)] =
    input.linesIterator.map:
      case s"$dir $n" =>
        val d = Dir.parse(dir)
        (d, n.toInt)
    .toList

  def part1(input: List[(Dir, Int)]): Int =
    val (horz, depth) =
      input.foldLeft((0, 0)):
        case ((horz, depth), (dir, n)) =>
          dir match
            case Dir.Forward => (horz + n, depth)
            case Dir.Down    => (horz, depth + n)
            case Dir.Up      => (horz, depth - n)

    depth * horz

  def part2(input: List[(Dir, Int)]): Int =
    @tailrec
    def go(rest: List[(Dir, Int)], aim: Int, depth: Int, x: Int): (Int, Int) =
      rest match
        case Nil          => (x, depth)
        case head :: next =>
          head match
            case (Dir.Forward, i) => go(next, aim, depth + aim * i, x + i)
            case (Dir.Down, i)    => go(next, aim + i, depth, x)
            case (Dir.Up, i)      => go(next, aim - i, depth, x)
    val (x, depth) = go(input, 0, 0, 0)
    x * depth
